### PR TITLE
v2.0.0

### LINKED ISSUE #37: Support logical types
The Avro specification and `avsc` implementation support defining [Logical Types](https://avro.apache.org/docs/1.8.2/spec.html#Logical+Types). In `avsc` this is done by specifying a value for the `logicalTypes` field in `avro.Type.forSchema/2`'s options argument. We should support this feature  as it is a common use-case for production applications.

For example:
```ts
const type = avro.Type.forSchema(schema, { logicalTypes: { decimal: DecimalType } })
```

### PR DESCRIPTION
**Pull request by Nevon**
This PR is mainly a documentation update to reflect the API changes following #93. It also fixes #37.

This version is almost entirely non-breaking, but the migration guide points out a few things that have changed.

### CODE REVIEW

**Comment by Nevon** [comment]
I'm not too happy about the signature of `SchemaRegistry.register`.

```ts
interface Opts {
  subject?: string
}

public async register(schema: RawAvroSchema | ConfluentSchema, userOpts?: Opts): Promise<RegisteredSchema>
```

The problem with this signature is that if the `schema` type is Protobuf or JSON, `userOpts` isn't actually optional. It'll throw at runtime because we don't have a way to make up a subject for those protocols, so we need the user to supply `userOpts.subject`. For avro, however, it is optional because we create a subject if one is not provided.

I tried solving this with overloads, but it doesn't quite work:

```ts
export interface AvroConfluentSchema {
  type: SchemaType.AVRO
  schema: string
}

export interface ProtoConfluentSchema {
  type: SchemaType.PROTOBUF,
  schema: string
}

export interface JsonConfluentSchema {
  type: SchemaType.JSON
  schema: string
}

export type ConfluentSchema = AvroConfluentSchema | ProtoConfluentSchema | JsonConfluentSchema

public async register(
    schema: Exclude<ConfluentSchema, AvroConfluentSchema>,
    userOpts: Omit<Opts, "subject"> & { subject: string },
  ): Promise<RegisteredSchema>
  public async register(schema: RawAvroSchema | AvroConfluentSchema, userOpts?: Opts): Promise<RegisteredSchema>
  public async register(schema: RawAvroSchema | ConfluentSchema, userOpts?: Opts): Promise<RegisteredSchema> {}
```

Here there are two signatures available to the users:

1. If the schema is anything other than an Avro schema, `userOpts` is required and the `subject` property of the `Opts` is required.
2. If the schema is either a `RawAvroSchema` (the legacy type) or an `AvroConfluentSchema`, `userOpts` is optional.

The final signature is the implementation signature, which is not visible to users, and it accepts any type of schema and `userOpts` is optional. I.e. it is compatible with all the overloads.

This actually compiles just fine, but then in our tests we loop over all the schema types and run the same tests on all of them, except with different data, and I think Typescript complains because it only knows that it's a `ConfluentSchema` type, but not which concrete type the schema actually is:

```ts
const types = Object.keys(schemaStringsByType).map(str => SchemaType[str]) as Exclude<SchemaType, SchemaType.UNKNOWN>

types.forEach(type => {
  describe(`${type}`, () => {
  const subject: ConfluentSubject = {
    name: [type, 'com.org.domain.fixtures', 'AnotherPerson'].join('.'),
  }
  const schema: ConfluentSchema = {
    type,
    schema: schemaStringsByType[type].v1,
  }

  beforeEach(async () => {
    schemaRegistry = new SchemaRegistry(schemaRegistryArgs)

    // Complains here, because `schema` is `ConfluentSchema`, which doesn't match any of the overloads
    await schemaRegistry.register(schema, { subject: subject.name })
  })
})
```

The specific error I'm getting is:

```
No overload matches this call.
  Overload 1 of 2, '(schema: ProtoConfluentSchema | JsonConfluentSchema, userOpts: Pick<Opts, "compatibility" | "separator"> & { subject: string; }): Promise<...>', gave the following error.
    Argument of type 'ConfluentSchema' is not assignable to parameter of type 'ProtoConfluentSchema | JsonConfluentSchema'.
      Type 'AvroConfluentSchema' is not assignable to type 'ProtoConfluentSchema | JsonConfluentSchema'.
        Type 'AvroConfluentSchema' is not assignable to type 'JsonConfluentSchema'.
          Types of property 'type' are incompatible.
            Type 'SchemaType.AVRO' is not assignable to type 'SchemaType.JSON'.
  Overload 2 of 2, '(schema: AvroConfluentSchema | RawAvroSchema, userOpts?: Opts): Promise<RegisteredSchema>', gave the following error.
    Argument of type 'ConfluentSchema' is not assignable to parameter of type 'AvroConfluentSchema | RawAvroSchema'.
      Type 'ProtoConfluentSchema' is not assignable to type 'AvroConfluentSchema | RawAvroSchema'.
        Type 'ProtoConfluentSchema' is not assignable to type 'AvroConfluentSchema'.
          Types of property 'type' are incompatible.
            Type 'SchemaType.PROTOBUF' is not assignable to type 'SchemaType.AVRO'.ts(2769)
SchemaRegistry.ts(89, 16): The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.
```

**Comment by Nevon** [comment]
Fixed the above mentioned problem. Of course, Typescript was right and I'm an idiot. If you don't know what concrete type you have in the union (like in the loop), the only safe thing is to require the options to be there, which there was no exposed overload for.

File: docs/v2.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+Since we now support multiple schema types, these options have been moved
+one level into a schema type specific option:
+
+```js
+const { SchemaRegistry, SchemaType } = require('@kafkajs/confluent-schema-registry')
+
+const registry = new SchemaRegistry(
+  { host: 'http://localhost:8081' },
+  {
+    [SchemaType.AVRO]: { forSchemaOptions: { noAnonymousTypes: true }},
```
Comment (malkiz-rapid) [inline docs/v2.md:101]: I think this should be `[SchemaType.AVRO]: { noAnonymousTypes: true }` (remove the `forSchemaOptions`)

File: src/SchemaRegistry.ts [hunk 2]
```diff
@@ -33,7 +34,7 @@ interface RegisteredSchema {
 interface Opts {
   compatibility?: COMPATIBILITY
   separator?: string
-  subject?: string
+  subject: string
```
Comment (malkiz-rapid) [inline src/SchemaRegistry.ts:37]: is this change necessary? wouldn't it be backwards incompatible for TS users who use Avro without specifying a subject?

File: src/SchemaRegistry.ts [hunk 3]
```diff
@@ -57,23 +58,41 @@ export default class SchemaRegistry {
...
     return confluentSchema
   }
 
+  public async register(
+    schema: Exclude<ConfluentSchema, AvroConfluentSchema>,
+    userOpts: Opts,
+  ): Promise<RegisteredSchema>
+  public async register(
+    schema: RawAvroSchema | AvroConfluentSchema,
+    userOpts?: Omit<Opts, 'subject'> & { subject?: string },
```
Comment (malkiz-rapid) [inline src/SchemaRegistry.ts:90]: I'm not sure what's going on here with these multiple signatures.
I assume that you wanted to indicate that only Avro schemas with the v1 api don't have to specify a subject.
But then why does the 3rd signature below also have `RawAvroSchema`?
It would make more sense to me if we had 1 signature to define the old v1 api and 1 signature for the new v2 api. But maybe I just don't understand what you tried to do here.

Comment (malkiz-rapid) [reply src/SchemaRegistry.ts:90]: nevermind, I have just now read your comments above. you can ignore this comment :)

File: docs/v2.md (hunk 1 above)
Comment (Nevon) [reply docs/v2.md]: Fixed in c5b3134 👍

**Comment by Nevon** [comment]
One thing I haven't tried is what happens if you build this and try to use it from Typescript. Primarily what I'm unsure of is if we actually export any of the types like `SchemaType` for example. Otherwise I should either make sure to re-export that from the index or change all the examples and documentation to use string literals instead of the enums.

**Review by tulios — APPROVED** [review]
